// ionic/dist/commands/deploy/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_framework_1 = require("@ionic/cli-framework");
const namespace_1 = require("@ionic/cli-utils/lib/namespace");
class DeployNamespace extends namespace_1.Namespace {
    getMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                name: 'deploy',
                summary: 'Commands for working with Ionic Deploy',
                groups: [cli_framework_1.NamespaceGroup.Hidden],
                description: ` These commands make it easy to do all things deploy `,
            };
        });
    }
    getCommands() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new namespace_1.CommandMap([
                ['manifest', () => tslib_1.__awaiter(this, void 0, void 0, function* () { const { DeployManifestCommand } = yield Promise.resolve().then(() => require('./manifest')); return new DeployManifestCommand(this); })],
            ]);
        });
    }
}
exports.DeployNamespace = DeployNamespace;
