// ionic/dist/commands/signup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
class SignupCommand extends command_1.Command {
    getMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                name: 'signup',
                type: 'global',
                summary: 'Create an account for Ionic Pro',
                description: `
Learn more about Ionic Pro: ${chalk_1.default.bold('https://ionicframework.com/pro')}

If you are having issues signing up, please get in touch with our Support${chalk_1.default.cyan('[1]')}.

${chalk_1.default.cyan('[1]')}: ${chalk_1.default.bold('https://ionicframework.com/support/request')}
      `,
            };
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const opn = yield Promise.resolve().then(() => require('opn'));
            const dashUrl = this.env.config.getDashUrl();
            yield opn(`${dashUrl}/signup?source=cli`, { wait: false });
            this.env.log.ok('Launched signup form in your browser!');
        });
    }
}
exports.SignupCommand = SignupCommand;
